Added Niklaus Wirth formulation of quicksort with some added twists. …

olehjalmar opened the pull request:
…Changed makefile to unroll loops (Yes, it does make a difference).

This version of quicksort has been my standby sorting algorithm for years. It is a bit more simplistic than the version already in sort.h, but performs well, especially with loop unrolling enabled.

swenson commented:
This definitely looks interesting! Thanks for the contribution.

A few questions and requests:

* How does this compare to the normal quick sort? (And what kind of computer are you checking it on?)
* Could you run `make format` so that the code will conform?
* It looks like there are some things in here that strict C89 compilers won't process, like mixed variable declarations and initializations. (I mess this up constantly.) I need to find the right flags or make a test I can run to catch these :)

I'll download this branch and see if I can benchmark. I've been meaning to make some sort of benchmark suite so that we can evaluate these sorts of changes and see their impact on different platforms.

swenson commented:
Oh, also, we already have binary insertion sort as a function -- would it make sense to call that instead of having the vanilla insertion sort macro?

swenson commented:
I just pushed a commit to the master branch that contains a benchmark utility: https://github.com/swenson/sort/blob/master/benchmark.c

You should be able to run before/after numbers and compare them with Go's `benchcmp` tool.

I can do this in the next day or so for a few architectures to see how this compares.

It looks like you'll need to rebase or merge with the new master, as I made changes to the `Makefile`. Sorry about that.

olehjalmar commented:
No problem, make any changes you need to.
I fixed a bug today, btw., which was introduced when I switched from signed
to unsigned index type, so you may want to pull the head of my branch.

Ole-Hj.

On Wed, Nov 8, 2017 at 5:10 AM, Christopher Swenson <
notifications@github.com> wrote:

> I just pushed a commit to the master branch that contains a benchmark
> utility: https://github.com/swenson/sort/blob/master/benchmark.c
>
> You should be able to run before/after numbers and compare them with Go's
> benchcmp tool.
>
> I can do this in the next day or so for a few architectures to see how
> this compares.
>
> It looks like you'll need to rebase or merge with the new master, as I
> made changes to the Makefile. Sorry about that.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/swenson/sort/pull/43#issuecomment-342705452>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Af6fch36RIQMv9TOtV6XEjvK_gBvlYdqks5s0SmagaJpZM4QVSUe>
> .
>

olehjalmar commented:
Yes, that should not be a problem. I just never bothered to implement a
binary insertion sort since I use it to sort just a few elements, but it
could be interesting to see if it changes the timing.

On Tue, Nov 7, 2017 at 10:32 PM, Christopher Swenson <
notifications@github.com> wrote:

> Oh, also, we already have binary insertion sort as a function -- would it
> make sense to call that instead of having the vanilla insertion sort macro?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/swenson/sort/pull/43#issuecomment-342629305>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Af6fcoIJVhZbTm7cAcNGqeIQu2L4IqIZks5s0Mx7gaJpZM4QVSUe>
> .
>

olehjalmar commented:
Are you thinking of Jon Bentley's version? Compared to that, both Wirth's
and the Wikipedia version are bloated...

void quicksort(int l, int u)
{   int i, m;
    if (l >= u) return;
    swap(l, randint(l, u));
    m = l;
    for (i = l+1; i < = u; i++)
        if (x[i] < x[l])
            swap(++m, i);
    swap(l, m);
    quicksort(l, m-1);
    quicksort(m+1, u);

}


It's very elegant, but Wirth's version is usually a bit faster. I don't
know if there is any standard quicksort, really, except the basic algorithm
of swapping elements which are on the wrong side. I kind of stuck with
Wirth's version after a series of experiments when I was trying to
duplicate the results from Jim Gray's Alphasort back about 1990.  I have
tried with a random pivot element, like Bentley does above, and it makes it
a bit slower on average, but I would assume it avoids some pathological
cases. Another alternative is to use a median like the quicksort code
already in sort.h.

I did run make format, but there may be limits to what it is willing to
change. I could run my code through clang-format first, maybe, then make
format.

I cannot see any mixed declarations and code, remember that even in K&R C
it is legal to declare variables at the beginning of any new scope.

On Tue, Nov 7, 2017 at 10:31 PM, Christopher Swenson <
notifications@github.com> wrote:

> This definitely looks interesting! Thanks for the contribution.
>
> A few questions and requests:
>
>    - How does this compare to the normal quick sort? (And what kind of
>    computer are you checking it on?)
>    - Could you run make format so that the code will conform?
>    - It looks like there are some things in here that strict C89
>    compilers won't process, like mixed variable declarations and
>    initializations. (I mess this up constantly.) I need to find the right
>    flags or make a test I can run to catch these :)
>
> I'll download this branch and see if I can benchmark. I've been meaning to
> make some sort of benchmark suite so that we can evaluate these sorts of
> changes and see their impact on different platforms.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/swenson/sort/pull/43#issuecomment-342628932>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Af6fcqJ51f53vBgRpCer5hiug5HRlIDGks5s0MwngaJpZM4QVSUe>
> .
>

olehjalmar commented:
My computer is a Toshiba Tecra laptop running Linux Mint.

Linux ole-TECRA-R940 4.4.0-53-generic #74-Ubuntu SMP Fri Dec 2 15:59:10 UTC
2016 x86_64 x86_64 x86_64 GNU/Linux

ole@ole-TECRA-R940 ~/sort $ cat /proc/cpuinfo
processor : 0
vendor_id : GenuineIntel
cpu family : 6
model : 58
model name : Intel(R) Core(TM) i5-3340M CPU @ 2.70GHz
stepping : 9
microcode : 0x15
cpu MHz : 1210.042
cache size : 3072 KB
physical id : 0
siblings : 4
core id : 0
cpu cores : 2
apicid : 0
initial apicid : 0
fpu : yes
fpu_exception : yes
cpuid level : 13
wp : yes
flags : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat
pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx rdtscp lm
constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc
aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2
ssse3 cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer
aes xsave avx f16c rdrand lahf_lm epb tpr_shadow vnmi flexpriority ept vpid
fsgsbase smep erms xsaveopt dtherm ida arat pln pts
bugs :
bogomips : 5387.89
clflush size : 64
cache_alignment : 64
address sizes : 36 bits physical, 48 bits virtual
power management:

processor : 1
vendor_id : GenuineIntel
cpu family : 6
model : 58
model name : Intel(R) Core(TM) i5-3340M CPU @ 2.70GHz
stepping : 9
microcode : 0x15
cpu MHz : 1202.132
cache size : 3072 KB
physical id : 0
siblings : 4
core id : 0
cpu cores : 2
apicid : 1
initial apicid : 1
fpu : yes
fpu_exception : yes
cpuid level : 13
wp : yes
flags : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat
pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx rdtscp lm
constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc
aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2
ssse3 cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer
aes xsave avx f16c rdrand lahf_lm epb tpr_shadow vnmi flexpriority ept vpid
fsgsbase smep erms xsaveopt dtherm ida arat pln pts
bugs :
bogomips : 5387.89
clflush size : 64
cache_alignment : 64
address sizes : 36 bits physical, 48 bits virtual
power management:

processor : 2
vendor_id : GenuineIntel
cpu family : 6
model : 58
model name : Intel(R) Core(TM) i5-3340M CPU @ 2.70GHz
stepping : 9
microcode : 0x15
cpu MHz : 1211.730
cache size : 3072 KB
physical id : 0
siblings : 4
core id : 1
cpu cores : 2
apicid : 2
initial apicid : 2
fpu : yes
fpu_exception : yes
cpuid level : 13
wp : yes
flags : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat
pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx rdtscp lm
constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc
aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2
ssse3 cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer
aes xsave avx f16c rdrand lahf_lm epb tpr_shadow vnmi flexpriority ept vpid
fsgsbase smep erms xsaveopt dtherm ida arat pln pts
bugs :
bogomips : 5387.89
clflush size : 64
cache_alignment : 64
address sizes : 36 bits physical, 48 bits virtual
power management:

processor : 3
vendor_id : GenuineIntel
cpu family : 6
model : 58
model name : Intel(R) Core(TM) i5-3340M CPU @ 2.70GHz
stepping : 9
microcode : 0x15
cpu MHz : 1201.394
cache size : 3072 KB
physical id : 0
siblings : 4
core id : 1
cpu cores : 2
apicid : 3
initial apicid : 3
fpu : yes
fpu_exception : yes
cpuid level : 13
wp : yes
flags : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat
pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx rdtscp lm
constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc
aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2
ssse3 cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer
aes xsave avx f16c rdrand lahf_lm epb tpr_shadow vnmi flexpriority ept vpid
fsgsbase smep erms xsaveopt dtherm ida arat pln pts
bugs :
bogomips : 5387.89
clflush size : 64
cache_alignment : 64
address sizes : 36 bits physical, 48 bits virtual
power management:


On Tue, Nov 7, 2017 at 10:31 PM, Christopher Swenson <
notifications@github.com> wrote:

> This definitely looks interesting! Thanks for the contribution.
>
> A few questions and requests:
>
>    - How does this compare to the normal quick sort? (And what kind of
>    computer are you checking it on?)
>    - Could you run make format so that the code will conform?
>    - It looks like there are some things in here that strict C89
>    compilers won't process, like mixed variable declarations and
>    initializations. (I mess this up constantly.) I need to find the right
>    flags or make a test I can run to catch these :)
>
> I'll download this branch and see if I can benchmark. I've been meaning to
> make some sort of benchmark suite so that we can evaluate these sorts of
> changes and see their impact on different platforms.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/swenson/sort/pull/43#issuecomment-342628932>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Af6fcqJ51f53vBgRpCer5hiug5HRlIDGks5s0MwngaJpZM4QVSUe>
> .
>

olehjalmar commented:
First, two runs with binary insertion sort

./benchmark | tee benchmark.txt
BenchmarkQsort100000_x86_64 86 11721488.4 ns/op
BenchmarkBinary_insertion_sort100000_x86_64 1 1505905000.0 ns/op
BenchmarkQuick_sort100000_x86_64 161 6216720.5 ns/op
BenchmarkWirth_quick_sort100000_x86_64 146 6859075.3 ns/op
BenchmarkMerge_sort100000_x86_64 117 8602914.5 ns/op
BenchmarkHeap_sort100000_x86_64 108 9282629.6 ns/op
BenchmarkShell_sort100000_x86_64 95 10574221.1 ns/op
BenchmarkTim_sort100000_x86_64 108 9267324.1 ns/op
BenchmarkMerge_sort_in_place100000_x86_64 129 7757248.1 ns/op
BenchmarkGrail_sort100000_x86_64 103 9732262.1 ns/op
BenchmarkSqrt_sort100000_x86_64 122 8242008.2 ns/op
BenchmarkRec_stable_sort100000_x86_64 31 32881806.5 ns/op
BenchmarkGrail_sort_dyn_buffer100000_x86_64 110 9092200.0 ns/op
./stresstest
       selection sort -- stable
binary insertion sort -- stable
          bubble sort -- stable
           quick sort -- UNSTABLE
           merge sort -- stable
            heap sort -- UNSTABLE
           shell sort -- UNSTABLE
             tim sort -- stable
merge (in-place) sort -- UNSTABLE
           grail sort -- stable
            sqrt sort -- stable
      rec stable sort -- stable
grail sort dyn byffer -- stable
-------
Running tests with random numbers:
-------
stdlib qsort                  - ok,    58335.0 usec
sort.h selection_sort         - ok,   573509.0 usec
sort.h bubble_sort            - ok,   602089.0 usec
sort.h binary_insertion_sort  - ok,   171768.0 usec
sort.h quick_sort             - ok,    31164.0 usec
sort.h wirth_quick_sort       - ok,    37056.0 usec
sort.h merge_sort             - ok,    45597.0 usec
sort.h heap_sort              - ok,    42063.0 usec
sort.h shell_sort             - ok,    45305.0 usec
sort.h tim_sort               - ok,    46303.0 usec
sort.h merge_sort_in_place    - ok,    46896.0 usec
sort.h grail_sort             - ok,    61282.0 usec
sort.h sqrt_sort              - ok,    39993.0 usec
sort.h rec_stable_sort        - ok,   131309.0 usec
sort.h grail_sort_dyn_buffer  - ok,    56822.0 usec
-------
Running tests with same number:
-------
stdlib qsort                  - ok,    17687.0 usec
sort.h selection_sort         - ok,   263255.0 usec
sort.h bubble_sort            - ok,      525.0 usec
sort.h binary_insertion_sort  - ok,      566.0 usec
sort.h quick_sort             - ok,      580.0 usec
sort.h wirth_quick_sort       - ok,      516.0 usec
sort.h merge_sort             - ok,    16709.0 usec
sort.h heap_sort              - ok,     3010.0 usec
sort.h shell_sort             - ok,     7022.0 usec
sort.h tim_sort               - ok,      724.0 usec
sort.h merge_sort_in_place    - ok,     1733.0 usec
sort.h grail_sort             - ok,    11049.0 usec
sort.h sqrt_sort              - ok,    13552.0 usec
sort.h rec_stable_sort        - ok,    11666.0 usec
sort.h grail_sort_dyn_buffer  - ok,    11124.0 usec
-------
Running tests with sorted numbers:
-------
stdlib qsort                  - ok,    17989.0 usec
sort.h selection_sort         - ok,   265472.0 usec
sort.h bubble_sort            - ok,      674.0 usec
sort.h binary_insertion_sort  - ok,      783.0 usec
sort.h quick_sort             - ok,     6923.0 usec
sort.h wirth_quick_sort       - ok,      606.0 usec
sort.h merge_sort             - ok,    16983.0 usec
sort.h heap_sort              - ok,    34962.0 usec
sort.h shell_sort             - ok,     6852.0 usec
sort.h tim_sort               - ok,      666.0 usec
sort.h merge_sort_in_place    - ok,     1865.0 usec
sort.h grail_sort             - ok,    21229.0 usec
sort.h sqrt_sort              - ok,    12553.0 usec
sort.h rec_stable_sort        - ok,    19237.0 usec
sort.h grail_sort_dyn_buffer  - ok,    16554.0 usec
-------
Running tests with sorted blocks of length 10:
-------
stdlib qsort                  - ok,    50699.0 usec
sort.h selection_sort         - ok,   522094.0 usec
sort.h bubble_sort            - ok,   548615.0 usec
sort.h binary_insertion_sort  - ok,   168557.0 usec
sort.h quick_sort             - ok,    29696.0 usec
sort.h wirth_quick_sort       - ok,    35552.0 usec
sort.h merge_sort             - ok,    39137.0 usec
sort.h heap_sort              - ok,    43315.0 usec
sort.h shell_sort             - ok,    41454.0 usec
sort.h tim_sort               - ok,    40751.0 usec
sort.h merge_sort_in_place    - ok,    37253.0 usec
sort.h grail_sort             - ok,    56269.0 usec
sort.h sqrt_sort              - ok,    35740.0 usec
sort.h rec_stable_sort        - ok,   122194.0 usec
sort.h grail_sort_dyn_buffer  - ok,    52152.0 usec
-------
Running tests with sorted blocks of length 100:
-------
stdlib qsort                  - ok,    33814.0 usec
sort.h selection_sort         - ok,   451471.0 usec
sort.h bubble_sort            - ok,   486590.0 usec
sort.h binary_insertion_sort  - ok,   150528.0 usec
sort.h quick_sort             - ok,    25069.0 usec
sort.h wirth_quick_sort       - ok,    30020.0 usec
sort.h merge_sort             - ok,    27174.0 usec
sort.h heap_sort              - ok,    39813.0 usec
sort.h shell_sort             - ok,    30952.0 usec
sort.h tim_sort               - ok,    13429.0 usec
sort.h merge_sort_in_place    - ok,    25707.0 usec
sort.h grail_sort             - ok,    43812.0 usec
sort.h sqrt_sort              - ok,    26325.0 usec
sort.h rec_stable_sort        - ok,   114865.0 usec
sort.h grail_sort_dyn_buffer  - ok,    38883.0 usec
-------
Running tests with sorted blocks of length 10000:
-------
stdlib qsort                  - ok,    17102.0 usec
sort.h selection_sort         - ok,   266296.0 usec
sort.h bubble_sort            - ok,      510.0 usec
sort.h binary_insertion_sort  - ok,      559.0 usec
sort.h quick_sort             - ok,     6263.0 usec
sort.h wirth_quick_sort       - ok,      520.0 usec
sort.h merge_sort             - ok,    15741.0 usec
sort.h heap_sort              - ok,    34356.0 usec
sort.h shell_sort             - ok,     6357.0 usec
sort.h tim_sort               - ok,      625.0 usec
sort.h merge_sort_in_place    - ok,     1610.0 usec
sort.h grail_sort             - ok,    21175.0 usec
sort.h sqrt_sort              - ok,    12283.0 usec
sort.h rec_stable_sort        - ok,    18928.0 usec
sort.h grail_sort_dyn_buffer  - ok,    16154.0 usec
-------
Running tests with swapped size/2 pairs:
-------
stdlib qsort                  - ok,    18379.0 usec
sort.h selection_sort         - ok,   268747.0 usec
sort.h bubble_sort            - ok,    87857.0 usec
sort.h binary_insertion_sort  - ok,     3586.0 usec
sort.h quick_sort             - ok,     6901.0 usec
sort.h wirth_quick_sort       - ok,     1516.0 usec
sort.h merge_sort             - ok,    16624.0 usec
sort.h heap_sort              - ok,    35054.0 usec
sort.h shell_sort             - ok,     8550.0 usec
sort.h tim_sort               - ok,     1942.0 usec
sort.h merge_sort_in_place    - ok,     4294.0 usec
sort.h grail_sort             - ok,    22910.0 usec
sort.h sqrt_sort              - ok,    13322.0 usec
sort.h rec_stable_sort        - ok,    20880.0 usec
sort.h grail_sort_dyn_buffer  - ok,    17674.0 usec
-------
Running tests with swapped size/8 pairs:
-------
stdlib qsort                  - ok,    18109.0 usec
sort.h selection_sort         - ok,   272512.0 usec
sort.h bubble_sort            - ok,    92416.0 usec
sort.h binary_insertion_sort  - ok,     3546.0 usec
sort.h quick_sort             - ok,     6828.0 usec
sort.h wirth_quick_sort       - ok,     1615.0 usec
sort.h merge_sort             - ok,    16895.0 usec
sort.h heap_sort              - ok,    35237.0 usec
sort.h shell_sort             - ok,     8521.0 usec
sort.h tim_sort               - ok,     2042.0 usec
sort.h merge_sort_in_place    - ok,     4311.0 usec
sort.h grail_sort             - ok,    23028.0 usec
sort.h sqrt_sort              - ok,    13429.0 usec
sort.h rec_stable_sort        - ok,    21286.0 usec
sort.h grail_sort_dyn_buffer  - ok,    17476.0 usec
ole@ole-TECRA-R940 ~/sort $ make
./benchmark | tee benchmark.txt
BenchmarkQsort100000_x86_64 86 11688767.4 ns/op
BenchmarkBinary_insertion_sort100000_x86_64 1 1504869000.0 ns/op
BenchmarkQuick_sort100000_x86_64 160 6272981.2 ns/op
BenchmarkWirth_quick_sort100000_x86_64 146 6875719.2 ns/op
BenchmarkMerge_sort100000_x86_64 117 8593572.6 ns/op
BenchmarkHeap_sort100000_x86_64 109 9221458.7 ns/op
BenchmarkShell_sort100000_x86_64 95 10636800.0 ns/op
BenchmarkTim_sort100000_x86_64 108 9326555.6 ns/op
BenchmarkMerge_sort_in_place100000_x86_64 129 7761589.1 ns/op
BenchmarkGrail_sort100000_x86_64 104 9673394.2 ns/op
BenchmarkSqrt_sort100000_x86_64 122 8204434.4 ns/op
BenchmarkRec_stable_sort100000_x86_64 31 33101032.3 ns/op
BenchmarkGrail_sort_dyn_buffer100000_x86_64 110 9118818.2 ns/op
./stresstest
       selection sort -- stable
binary insertion sort -- stable
          bubble sort -- stable
           quick sort -- UNSTABLE
           merge sort -- stable
            heap sort -- UNSTABLE
           shell sort -- UNSTABLE
             tim sort -- stable
merge (in-place) sort -- UNSTABLE
           grail sort -- stable
            sqrt sort -- stable
      rec stable sort -- stable
grail sort dyn byffer -- stable
-------
Running tests with random numbers:
-------
stdlib qsort                  - ok,    60758.0 usec
sort.h selection_sort         - ok,   574513.0 usec
sort.h bubble_sort            - ok,   604156.0 usec
sort.h binary_insertion_sort  - ok,   171318.0 usec
sort.h quick_sort             - ok,    31994.0 usec
sort.h wirth_quick_sort       - ok,    37507.0 usec
sort.h merge_sort             - ok,    47418.0 usec
sort.h heap_sort              - ok,    42306.0 usec
sort.h shell_sort             - ok,    43965.0 usec
sort.h tim_sort               - ok,    46116.0 usec
sort.h merge_sort_in_place    - ok,    42644.0 usec
sort.h grail_sort             - ok,    61165.0 usec
sort.h sqrt_sort              - ok,    41369.0 usec
sort.h rec_stable_sort        - ok,   131522.0 usec
sort.h grail_sort_dyn_buffer  - ok,    57355.0 usec
-------
Running tests with same number:
-------
stdlib qsort                  - ok,    17701.0 usec
sort.h selection_sort         - ok,   263885.0 usec
sort.h bubble_sort            - ok,      736.0 usec
sort.h binary_insertion_sort  - ok,      605.0 usec
sort.h quick_sort             - ok,      645.0 usec
sort.h wirth_quick_sort       - ok,      605.0 usec
sort.h merge_sort             - ok,    16897.0 usec
sort.h heap_sort              - ok,     2886.0 usec
sort.h shell_sort             - ok,     6878.0 usec
sort.h tim_sort               - ok,      794.0 usec
sort.h merge_sort_in_place    - ok,     1638.0 usec
sort.h grail_sort             - ok,    11261.0 usec
sort.h sqrt_sort              - ok,    13109.0 usec
sort.h rec_stable_sort        - ok,    11885.0 usec
sort.h grail_sort_dyn_buffer  - ok,    11103.0 usec
-------
Running tests with sorted numbers:
-------
stdlib qsort                  - ok,    18461.0 usec
sort.h selection_sort         - ok,   265805.0 usec
sort.h bubble_sort            - ok,      683.0 usec
sort.h binary_insertion_sort  - ok,      659.0 usec
sort.h quick_sort             - ok,     6655.0 usec
sort.h wirth_quick_sort       - ok,      593.0 usec
sort.h merge_sort             - ok,    16859.0 usec
sort.h heap_sort              - ok,    35043.0 usec
sort.h shell_sort             - ok,     7092.0 usec
sort.h tim_sort               - ok,      717.0 usec
sort.h merge_sort_in_place    - ok,     1586.0 usec
sort.h grail_sort             - ok,    21938.0 usec
sort.h sqrt_sort              - ok,    12609.0 usec
sort.h rec_stable_sort        - ok,    19684.0 usec
sort.h grail_sort_dyn_buffer  - ok,    16243.0 usec
-------
Running tests with sorted blocks of length 10:
-------
stdlib qsort                  - ok,    50551.0 usec
sort.h selection_sort         - ok,   616995.0 usec
sort.h bubble_sort            - ok,   549645.0 usec
sort.h binary_insertion_sort  - ok,   172686.0 usec
sort.h quick_sort             - ok,    31322.0 usec
sort.h wirth_quick_sort       - ok,    35708.0 usec
sort.h merge_sort             - ok,    40769.0 usec
sort.h heap_sort              - ok,    42418.0 usec
sort.h shell_sort             - ok,    42562.0 usec
sort.h tim_sort               - ok,    40713.0 usec
sort.h merge_sort_in_place    - ok,    36445.0 usec
sort.h grail_sort             - ok,    57649.0 usec
sort.h sqrt_sort              - ok,    36430.0 usec
sort.h rec_stable_sort        - ok,   123701.0 usec
sort.h grail_sort_dyn_buffer  - ok,    51777.0 usec
-------
Running tests with sorted blocks of length 100:
-------
stdlib qsort                  - ok,    34297.0 usec
sort.h selection_sort         - ok,   462116.0 usec
sort.h bubble_sort            - ok,   489917.0 usec
sort.h binary_insertion_sort  - ok,   152433.0 usec
sort.h quick_sort             - ok,    25555.0 usec
sort.h wirth_quick_sort       - ok,    30463.0 usec
sort.h merge_sort             - ok,    27650.0 usec
sort.h heap_sort              - ok,    40388.0 usec
sort.h shell_sort             - ok,    30662.0 usec
sort.h tim_sort               - ok,    13745.0 usec
sort.h merge_sort_in_place    - ok,    25483.0 usec
sort.h grail_sort             - ok,    43212.0 usec
sort.h sqrt_sort              - ok,    26186.0 usec
sort.h rec_stable_sort        - ok,    90228.0 usec
sort.h grail_sort_dyn_buffer  - ok,    38956.0 usec
-------
Running tests with sorted blocks of length 10000:
-------
stdlib qsort                  - ok,    17295.0 usec
sort.h selection_sort         - ok,   263567.0 usec
sort.h bubble_sort            - ok,      490.0 usec
sort.h binary_insertion_sort  - ok,      541.0 usec
sort.h quick_sort             - ok,     6288.0 usec
sort.h wirth_quick_sort       - ok,      519.0 usec
sort.h merge_sort             - ok,    15835.0 usec
sort.h heap_sort              - ok,    34714.0 usec
sort.h shell_sort             - ok,     6466.0 usec
sort.h tim_sort               - ok,      613.0 usec
sort.h merge_sort_in_place    - ok,     1547.0 usec
sort.h grail_sort             - ok,    21566.0 usec
sort.h sqrt_sort              - ok,    12235.0 usec
sort.h rec_stable_sort        - ok,    19085.0 usec
sort.h grail_sort_dyn_buffer  - ok,    16102.0 usec
-------
Running tests with swapped size/2 pairs:
-------
stdlib qsort                  - ok,    19217.0 usec
sort.h selection_sort         - ok,   271409.0 usec
sort.h bubble_sort            - ok,    91383.0 usec
sort.h binary_insertion_sort  - ok,     3926.0 usec
sort.h quick_sort             - ok,     7302.0 usec
sort.h wirth_quick_sort       - ok,     1612.0 usec
sort.h merge_sort             - ok,    17298.0 usec
sort.h heap_sort              - ok,    36230.0 usec
sort.h shell_sort             - ok,    10540.0 usec
sort.h tim_sort               - ok,     2208.0 usec
sort.h merge_sort_in_place    - ok,     4446.0 usec
sort.h grail_sort             - ok,    22438.0 usec
sort.h sqrt_sort              - ok,    13011.0 usec
sort.h rec_stable_sort        - ok,    20950.0 usec
sort.h grail_sort_dyn_buffer  - ok,    17505.0 usec
-------
Running tests with swapped size/8 pairs:
-------
stdlib qsort                  - ok,    18838.0 usec
sort.h selection_sort         - ok,   271515.0 usec
sort.h bubble_sort            - ok,    93877.0 usec
sort.h binary_insertion_sort  - ok,     3803.0 usec
sort.h quick_sort             - ok,     7105.0 usec
sort.h wirth_quick_sort       - ok,     1706.0 usec
sort.h merge_sort             - ok,    16953.0 usec
sort.h heap_sort              - ok,    36072.0 usec
sort.h shell_sort             - ok,     8640.0 usec
sort.h tim_sort               - ok,     2084.0 usec
sort.h merge_sort_in_place    - ok,     4514.0 usec
sort.h grail_sort             - ok,    23066.0 usec
sort.h sqrt_sort              - ok,    13695.0 usec
sort.h rec_stable_sort        - ok,    21789.0 usec
sort.h grail_sort_dyn_buffer  - ok,    17863.0 usec

Then two runs with straight insertion sort macro


./benchmark | tee benchmark.txt
BenchmarkQsort100000_x86_64 85 11846670.6 ns/op
BenchmarkBinary_insertion_sort100000_x86_64 1 1550059000.0 ns/op
BenchmarkQuick_sort100000_x86_64 158 6336132.9 ns/op
BenchmarkWirth_quick_sort100000_x86_64 158 6330936.7 ns/op
BenchmarkMerge_sort100000_x86_64 117 8608136.8 ns/op
BenchmarkHeap_sort100000_x86_64 107 9385579.4 ns/op
BenchmarkShell_sort100000_x86_64 93 10761064.5 ns/op
BenchmarkTim_sort100000_x86_64 114 8849140.4 ns/op
BenchmarkMerge_sort_in_place100000_x86_64 128 7872156.2 ns/op
BenchmarkGrail_sort100000_x86_64 102 9814529.4 ns/op
BenchmarkSqrt_sort100000_x86_64 122 8235442.6 ns/op
BenchmarkRec_stable_sort100000_x86_64 31 33028419.4 ns/op
BenchmarkGrail_sort_dyn_buffer100000_x86_64 110 9119245.5 ns/op
./stresstest
       selection sort -- stable
binary insertion sort -- stable
          bubble sort -- stable
           quick sort -- UNSTABLE
           merge sort -- stable
            heap sort -- UNSTABLE
           shell sort -- UNSTABLE
             tim sort -- stable
merge (in-place) sort -- UNSTABLE
           grail sort -- stable
            sqrt sort -- stable
      rec stable sort -- stable
grail sort dyn byffer -- stable
-------
Running tests with random numbers:
-------
stdlib qsort                  - ok,    78111.0 usec
sort.h selection_sort         - ok,  1076776.0 usec
sort.h bubble_sort            - ok,   645349.0 usec
sort.h binary_insertion_sort  - ok,   172218.0 usec
sort.h quick_sort             - ok,    31213.0 usec
sort.h wirth_quick_sort       - ok,    32749.0 usec
sort.h merge_sort             - ok,    44729.0 usec
sort.h heap_sort              - ok,    43357.0 usec
sort.h shell_sort             - ok,    44004.0 usec
sort.h tim_sort               - ok,    45557.0 usec
sort.h merge_sort_in_place    - ok,    40719.0 usec
sort.h grail_sort             - ok,    60636.0 usec
sort.h sqrt_sort              - ok,    43146.0 usec
sort.h rec_stable_sort        - ok,   132877.0 usec
sort.h grail_sort_dyn_buffer  - ok,    55801.0 usec
-------
Running tests with same number:
-------
stdlib qsort                  - ok,    17663.0 usec
sort.h selection_sort         - ok,   266090.0 usec
sort.h bubble_sort            - ok,      546.0 usec
sort.h binary_insertion_sort  - ok,      564.0 usec
sort.h quick_sort             - ok,      547.0 usec
sort.h wirth_quick_sort       - ok,      524.0 usec
sort.h merge_sort             - ok,    15820.0 usec
sort.h heap_sort              - ok,     2527.0 usec
sort.h shell_sort             - ok,     6796.0 usec
sort.h tim_sort               - ok,      564.0 usec
sort.h merge_sort_in_place    - ok,     1505.0 usec
sort.h grail_sort             - ok,    11376.0 usec
sort.h sqrt_sort              - ok,    14910.0 usec
sort.h rec_stable_sort        - ok,    12383.0 usec
sort.h grail_sort_dyn_buffer  - ok,    10683.0 usec
-------
Running tests with sorted numbers:
-------
stdlib qsort                  - ok,    17860.0 usec
sort.h selection_sort         - ok,   265985.0 usec
sort.h bubble_sort            - ok,      685.0 usec
sort.h binary_insertion_sort  - ok,      640.0 usec
sort.h quick_sort             - ok,     6676.0 usec
sort.h wirth_quick_sort       - ok,      632.0 usec
sort.h merge_sort             - ok,    16103.0 usec
sort.h heap_sort              - ok,    38073.0 usec
sort.h shell_sort             - ok,     6979.0 usec
sort.h tim_sort               - ok,      712.0 usec
sort.h merge_sort_in_place    - ok,     1603.0 usec
sort.h grail_sort             - ok,    21896.0 usec
sort.h sqrt_sort              - ok,    14227.0 usec
sort.h rec_stable_sort        - ok,    20672.0 usec
sort.h grail_sort_dyn_buffer  - ok,    16247.0 usec
-------
Running tests with sorted blocks of length 10:
-------
stdlib qsort                  - ok,    49295.0 usec
sort.h selection_sort         - ok,   840437.0 usec
sort.h bubble_sort            - ok,   594186.0 usec
sort.h binary_insertion_sort  - ok,   171027.0 usec
sort.h quick_sort             - ok,    30529.0 usec
sort.h wirth_quick_sort       - ok,    31707.0 usec
sort.h merge_sort             - ok,    38020.0 usec
sort.h heap_sort              - ok,    43072.0 usec
sort.h shell_sort             - ok,    40236.0 usec
sort.h tim_sort               - ok,    40493.0 usec
sort.h merge_sort_in_place    - ok,    35695.0 usec
sort.h grail_sort             - ok,    56480.0 usec
sort.h sqrt_sort              - ok,    39501.0 usec
sort.h rec_stable_sort        - ok,   124615.0 usec
sort.h grail_sort_dyn_buffer  - ok,    52794.0 usec
-------
Running tests with sorted blocks of length 100:
-------
stdlib qsort                  - ok,    34600.0 usec
sort.h selection_sort         - ok,   470543.0 usec
sort.h bubble_sort            - ok,   512525.0 usec
sort.h binary_insertion_sort  - ok,   151260.0 usec
sort.h quick_sort             - ok,    25056.0 usec
sort.h wirth_quick_sort       - ok,    25774.0 usec
sort.h merge_sort             - ok,    26423.0 usec
sort.h heap_sort              - ok,    40938.0 usec
sort.h shell_sort             - ok,    31801.0 usec
sort.h tim_sort               - ok,    13387.0 usec
sort.h merge_sort_in_place    - ok,    25536.0 usec
sort.h grail_sort             - ok,    46133.0 usec
sort.h sqrt_sort              - ok,    29347.0 usec
sort.h rec_stable_sort        - ok,    93233.0 usec
sort.h grail_sort_dyn_buffer  - ok,    38754.0 usec
-------
Running tests with sorted blocks of length 10000:
-------
stdlib qsort                  - ok,    17508.0 usec
sort.h selection_sort         - ok,   260871.0 usec
sort.h bubble_sort            - ok,      523.0 usec
sort.h binary_insertion_sort  - ok,      528.0 usec
sort.h quick_sort             - ok,     6247.0 usec
sort.h wirth_quick_sort       - ok,      513.0 usec
sort.h merge_sort             - ok,    15329.0 usec
sort.h heap_sort              - ok,    37890.0 usec
sort.h shell_sort             - ok,     6283.0 usec
sort.h tim_sort               - ok,      602.0 usec
sort.h merge_sort_in_place    - ok,     1547.0 usec
sort.h grail_sort             - ok,    21130.0 usec
sort.h sqrt_sort              - ok,    13960.0 usec
sort.h rec_stable_sort        - ok,    20462.0 usec
sort.h grail_sort_dyn_buffer  - ok,    16061.0 usec
-------
Running tests with swapped size/2 pairs:
-------
stdlib qsort                  - ok,    18525.0 usec
sort.h selection_sort         - ok,   270882.0 usec
sort.h bubble_sort            - ok,    91611.0 usec
sort.h binary_insertion_sort  - ok,     3721.0 usec
sort.h quick_sort             - ok,     7080.0 usec
sort.h wirth_quick_sort       - ok,     1619.0 usec
sort.h merge_sort             - ok,    16019.0 usec
sort.h heap_sort              - ok,    39012.0 usec
sort.h shell_sort             - ok,     8833.0 usec
sort.h tim_sort               - ok,     2064.0 usec
sort.h merge_sort_in_place    - ok,     4277.0 usec
sort.h grail_sort             - ok,    22626.0 usec
sort.h sqrt_sort              - ok,    14691.0 usec
sort.h rec_stable_sort        - ok,    22112.0 usec
sort.h grail_sort_dyn_buffer  - ok,    17787.0 usec
-------
Running tests with swapped size/8 pairs:
-------
stdlib qsort                  - ok,    18412.0 usec
sort.h selection_sort         - ok,   265424.0 usec
sort.h bubble_sort            - ok,    93292.0 usec
sort.h binary_insertion_sort  - ok,     3703.0 usec
sort.h quick_sort             - ok,     6940.0 usec
sort.h wirth_quick_sort       - ok,     1666.0 usec
sort.h merge_sort             - ok,    15994.0 usec
sort.h heap_sort              - ok,    38249.0 usec
sort.h shell_sort             - ok,     8831.0 usec
sort.h tim_sort               - ok,     2164.0 usec
sort.h merge_sort_in_place    - ok,     4379.0 usec
sort.h grail_sort             - ok,    23219.0 usec
sort.h sqrt_sort              - ok,    15143.0 usec
sort.h rec_stable_sort        - ok,    22395.0 usec
sort.h grail_sort_dyn_buffer  - ok,    17850.0 usec

It seems that the straight insertion sort is faster on my machine

On Tue, Nov 7, 2017 at 10:32 PM, Christopher Swenson <
notifications@github.com> wrote:

> Oh, also, we already have binary insertion sort as a function -- would it
> make sense to call that instead of having the vanilla insertion sort macro?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/swenson/sort/pull/43#issuecomment-342629305>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/Af6fcoIJVhZbTm7cAcNGqeIQu2L4IqIZks5s0Mx7gaJpZM4QVSUe>
> .
>

swenson commented:
Sorry I dropped the ball on this. Looking at the benchmark numbers, it looks like the Wirth implementation of quicksort is strictly worse than the implementation we already have, unless I am missing something?

swenson commented:
After some exhaustive testing as part of #63, I think Hoare partitioning algorithm (used here) does not seem to have any advantages over the current partitioning scheme, especially when combined with other optimizations in #63 (tail-call optimization, median-of-5, resorting to heap sort if the number of recursions is too high, and using bitonic sorting as the "small" sorting algorithm).
